<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

## Exercise 1 (Make me an odd list) ☻

In [5]:
oddsq_list = []
for number in range(0,10):
    if number % 2 == 0:
        continue
    else:
        oddsq_list += [number**2]
print(oddsq_list)

[1, 9, 25, 49, 81]


## Exercise 2 (Make me another odd list) ☻



In [1]:
[number**2 for number in range(11) if number %2 != 0 ]

[1, 9, 25, 49, 81]

## Exercise 3 (Time me!) ☻



In [4]:
%%timeit
oddsq_list = []
for number in range(0,10):
    if number % 2 == 0:
        continue
    else:
        oddsq_list += [number**2]
print(oddsq_list)

[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 4

In [3]:
%%timeit
[number**2 for number in range(11) if number %2 != 0 ]

689 ns ± 22.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


### Use the cell magic command %%timeit to time the previous solutions. Which of the two is faster?

list comprehension is faster with a mean execution time of 689 **nano** seconds as compared to the for loop, which has a mean execution time of 103 **micro** seconds. 

## Exercise 4 (A problem of decay) ☻



In [6]:
sample = 100
year = 0
while sample > 50:
    sample *= 0.95
    year += 1
print(year)

14


## Exercise 5 (Changes in CO) ☻

In [15]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]
years = [] # element is [year, change in concentration]
for i in range(len(co2_data)-1):
    if co2_data[i+1][1]-co2_data[i][1]>=3:
        add_year = [co2_data[i+1][0], co2_data[i+1][1]-co2_data[i][1]]
        years += add_year
    else:
        continue
print(years)

[2016, 3.410000000000025]
